### DISCUSSION TITLE
AbaqusUMATStress for different constant_properties for each element or grain

### DISCUSSION THREAD

**Discussion by LifengGan** [idaholab/moose#27809] [Q&A Modules: Solid mechanics] [answered]
Dear moose user,

How can I load `constant_properties`  using a list of data as the euler angle for each element or grain in `AbaqusUMATStress`.

Is it possible somehow to use the `PropertyReadFile` to do this?
```
[Materials]
  [umat]
    type = AbaqusUMATStress
    constant_properties = 'eulerang1 eulerang2 eulerang3' 
    plugin = '../../../plugins/crystal_plasticity'
    num_state_vars = 0
    temperature = temperature
    use_one_based_indexing = true
  []
[]
```

```
[UserObjects]
  [./prop_read]
    type = PropertyReadFile
    prop_file_name = 'euler_angle_file.txt'
    nprop = 3
    read_type = element
  [../]
[]
```

Best regards,
LifengGan

**Comment by dschwen** [comment, marked as answer]
Those would not be constant properties, but `external_fields` and you would have to put the Euler angles into a set of AuxVariables. If you want to go with `constant_properties` you'd have to block restrict the `AbaqusUMATStress` material and add it multiple times (once for each block) and supply it with different values for the Euler angles each time. If you have a small number of Euler angles this could be done by hand. For a larger number you could either generate an input using a Python script or write a MOOSE Action that adds the `AbaqusUMATStress` using data read from a file.
